tmt/steps/discover/__init__.py

```diff
@@ -146,55 +147,69 @@ def save(self) -> None:
         super().save()
 
         # Create tests.yaml with the full test data
-        tests = dict([
-            (test.name, test.export(format_='dict'))
-            for test in self.tests()])
-        self.write('tests.yaml', tmt.utils.dict_to_yaml(tests))
+        raw_test_data = {
+            test.name: test.export(format_='dict')
+            for test in self.tests()
+            }
+
+        self.write('tests.yaml', tmt.utils.dict_to_yaml(raw_test_data))
 
         # Create 'run.yaml' with the list of tests for the executor
         if not self.tests():
             return
-        tests = dict([
-            (test.name, test.export(format_='execute'))
-            for test in self.tests()])
-        self.write('run.yaml', tmt.utils.dict_to_yaml(tests, width=1000000))
+        raw_test_data = {
+            test.name: test.export(format_='execute')
+            for test in self.tests()
+            }
+        self.write('run.yaml', tmt.utils.dict_to_yaml(raw_test_data, width=1000000))
 
     def _discover_from_execute(self) -> None:
         """ Check the execute step for possible shell script tests """
 
         # Check scripts for command line and data, convert to list if needed
         scripts = self.plan.execute.opt('script')
         if not scripts:
-            scripts = self.plan.execute.data[0].get('script')
+            scripts = getattr(self.plan.execute.data[0], 'script', [])
         if not scripts:
             return
         if isinstance(scripts, str):
             scripts = [scripts]
 
+        # Avoid circular imports
+        import tmt.base
+        import tmt.steps.discover.shell
+
         # Give a warning when discover step defined as well
-        default_data = [{'name': 'default-0', 'how': 'shell'}]
-        if self.data and self.data != default_data:
+        if self.data and not all(datum.is_bare for datum in self.data):
             raise tmt.utils.DiscoverError(
                 "Use either 'discover' or 'execute' step "
                 "to define tests, but not both.")
 
+        # TODO: there is no mention of `duration` in **execute** step specification
```

Comment:
Wrong, there is. But it's probably slightly wrong:

> The DURATION is an optional section stating how long did the test run. Its value is in the hh:mm:ss format.

@psss isn't it supposed to be `1h` or `5m` format, instead of `01:00:00`?

It's probably missing from schema though, @happz standalone PR, please.

Reply:
The sleep format e.g. `5m` is used in the [duration](https://tmt.readthedocs.io/en/latest/spec/tests.html#duration) attribute of tests (maximum time test, to be written by users), the `01:00:00` format is used for `results.yaml` and is documented in the [execute](https://tmt.readthedocs.io/en/latest/spec/plans.html#execute) step spec (real time which test execution took, to be generated by automation). The `duration` key is also mentioned under [execute/script](https://tmt.readthedocs.io/en/latest/spec/plans.html#script).

Reply:
I see, that's exactly the answer I was looking for, thanks. I'll reorder the code a bit, now when I understand it a bit more. Leaving discussion open for later re-check.